SkyModelData

aelanman opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Replaces the use of recarrays for sharing SkyModel data with a new object, defined in `pyuvsim.simsetup`, that handles all the required behavior. Also cuts down on memory bloat due to repeating the `freq_array` Ncomponents times when storing as a column in a recarray.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

pyradiosky inherited the "recarray" form of sky models from `pyuvsim`, which was a hacky solution to share sky model data with all MPI processes in a container that is MPI-friendly. It also made sense with how votables and text files were initially read in (as numpy arrays). This partially saved on memory, because much of the data could be put in shared memory windows on each node. As pyradiosky evolved, however, the corresponding recarray definition got more complicated. The addition of the `freq_array` for instance, required that the recarray carry around a column that's just repetitions of the `freq_array`, which is horribly inefficient.

The main impediment to directly using SkyModels in MPI is that several attributes are defined as astropy Quantities (or derived classes, such as Angles, Latitudes, Longitudes), and this information is lost when sharing the data in MPI, unless the slow serialization methods are used (as in `mpi4py.MPI.bcast`). I've added a function to allow broadcasting astropy Quantities to shared memory, but haven't used it much.

This defines a new class, `SkyModelData`, which carries the SkyModel parameters as ndarrays. These arrays can be shared across MPI processes by running the SkyModelData.share() method, which puts arrays in shared memory (if that array has an axis of length `Ncomponents`).

Further, it minimizes Stokes information that is broadcast by only keeping non-stokes I flux values for components which have nonzero polarization. Most simulations won't have many polarized sources, so this can reduce required memory.

closes #293 

<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an replace the space with an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Reference simulation update or replacement
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the documentation to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [ ] I have checked that I reproduce the reference simulations or if there are differences they are explained below (if appropriate). If there are changes that are correct, I will update the reference simulation files after this PR is merged.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvsim/blob/master/CHANGELOG.md).

bhazelton commented on pyuvsim/simsetup.py:
```diff
@@ -285,29 +287,199 @@ def create_mock_catalog(time, arrangement='zenith', array_location=None, Nsrcs=N
+            if sky_in._spectral_index.required:
+                self.spectral_index = sky_in.spectral_index
+
+    def __getitem__(self, inds):
```
After reading some about `__getitem__` this seems to be a non-standard implementation -- it's returning a downselected copy of the object which is not what `__getitem__` is generally used for.

bhazelton commented on pyuvsim/simsetup.py:
```diff
@@ -285,29 +287,199 @@ def create_mock_catalog(time, arrangement='zenith', array_location=None, Nsrcs=N
+    spectral_index = None
+    polarized = None
+
+    ncomp_attrs = ['stokes_I', 'stokes_Q', 'stokes_U', 'stokes_V',
+                   'ra', 'dec', 'reference_frequency', 'spectral_index', 'name']
```
Is this supposed to list attributes that have length `Ncomponents`? I don't think stokes_Q/U/V generally do have that length, do they? I think that they have the length of the number of polarized sources. If that's the case, I'm concerned about where you use it in `__getitem__` to downselect using the component indices. On the other hand, you have a test that looks like it is testing that this works, so I'm confused.

If stokes_Q/U/V do have length `Ncomponents`, I don't see why to use this new object -- it will have the same size as the SkyModel object.

aelanman replied:
Using a new class helps because it keeps the data in numpy arrays with fixed units instead of in Angles/Latitudes/Quantity etc. It's a lot easier to handle with MPI this way.

You're right though, about the stokes_Q/U/V arrays... they don't have length `Ncomponents` unless every source is polarized. This does break __getitem__ so I'll have to fix that too. Thanks for pointing it out!

bhazelton commented on pyuvsim/uvsim.py:
```diff
@@ -273,17 +273,17 @@ def uvdata_to_task_iter(task_ids, input_uv, catalog, beam_list, beam_dict, Nsky_
 
     #   Skymodel will now be passed in as a catalog array.
-    if not isinstance(catalog, np.ndarray):
+    if not isinstance(catalog, SkyModelData):
         raise TypeError("catalog must be a record array")
```
I think this error message is wrong.

aelanman replied:
Thanks, it's fixed now.

aelanman commented:
@bhazelton I've renamed the `__getitem__` function to `subselect`, and changed its syntax to properly handle the polarization arrays. There's some other changes in there too to ensure that, under normal operation, the arrays are not copied (i.e., views are returned). I have a test in place to make sure that the arrays aren't copied, but to be safe I'm running a simulation with a large catalog to make sure it doesn't use more memory than previously.

The subselect method is, unfortunately, necessary for the (albeit, rare) case that Npus > Nbls * Ntimes * Nfreqs, but Nsrcs > Npus. In that case, the source axis is split across MPI processes, instead of in chunks on each rank.

bhazelton commented on pyuvsim/simsetup.py:
```diff
@@ -369,24 +574,24 @@ def initialize_catalog_from_params(obs_params, input_uv=None):
             catalog = os.path.join(param_dict['config_path'], catalog)
         if catalog.endswith("txt"):
-            catalog = pyradiosky.read_text_catalog(catalog, return_table=True)
+            sky = pyradiosky.read_text_catalog(catalog)
```
The function you're calling here is deprecated. Can you use `SkyModel.read_text_catalog`?

aelanman replied:
updated!

bhazelton commented on pyuvsim/simsetup.py:
```diff
@@ -369,24 +574,24 @@ def initialize_catalog_from_params(obs_params, input_uv=None):
                     spectral_type = "flat"
-                catalog = pyradiosky.skymodel.read_gleam_catalog(
-                    catalog, spectral_type=spectral_type, return_table=True
+                sky = pyradiosky.skymodel.read_gleam_catalog(
+                    catalog, spectral_type=spectral_type
```
The function you're calling here is deprecated. Can you use `SkyModel.read_gleam_catalog`?

bhazelton commented on pyuvsim/simsetup.py:
```diff
@@ -409,20 +614,23 @@ def initialize_catalog_from_params(obs_params, input_uv=None):
                 vo_params["reference_frequency"] = None
-                catalog = pyradiosky.read_votable_catalog(
-                    catalog, **vo_params, return_table=True
+                sky = pyradiosky.read_votable_catalog(
+                    catalog, **vo_params
```
The function you're calling here is deprecated. Can you use `SkyModel.read_votable_catalog`?

bhazelton commented on pyuvsim/simsetup.py:
```diff
@@ -409,20 +614,23 @@ def initialize_catalog_from_params(obs_params, input_uv=None):
+                    catalog, **vo_params
                 )
         elif catalog.endswith('hdf5'):
             hpmap, inds, freqs = pyradiosky.read_healpix_hdf5(catalog)
             sky = pyradiosky.healpix_to_sky(hpmap, inds, freqs)
```
These functions are deprecated. Can you use `SkyModel.read_healpix_hdf5`?

aelanman commented:
@bhazelton I've replaced the deprecated function calls, but I'd like to wait on this until we've finished making the stokes array a Quantity in pyradiosky.

aelanman commented:
@bhazelton This should be ready now. It's compatible with the new changes to pyradiosky and with the current release.

I have another branch which won't be compatible with the current release, so if possible we should make a new pyradiosky release soon.

aelanman commented:
Never mind... I just realized that `hera_sim` is relying on `simsetup.initialize_catalog_from_params` to return a recarray, so I'll need to add backwards compatibility.

We should probably add a github action to run the `hera_sim` tests, since it relies on pyuvsim so heavily. I'm not really sure how to do that though.

steven-murray commented:
@aelanman if there's a better way of using the `pyuvsim.simsetup` routines now (or directly through `pyradiosky`, I'd be happy to implement it in `hera_sim`!

aelanman commented:
@steven-murray The main function (`initialize_catalog_from_params`) hasn't changed much, except to return a `SkyModel` directly instead of a recarray. I've restored the option to return a recarray and made it the default for compatibility, but ultimately we don't want that to be the default behavior.

It made some sense for `initialize_catalog...` to return a recarray back when that was an intermediate step to reading in catalogs (astropy's VOTable reader and the numpy text parser both returned recarrays). That has been changed in pyradiosky, so the function was making a SkyModel and then converting it to recarray (presumably then to be converted back by the user...). This was only necessary for MPI, but this branch fully removes pyuvsim's need to convert catalogs to recarrays.
